api/types/instance.go

```diff
@@ -39,14 +41,57 @@ func (f InstanceFilter) Match(i Instance) bool {
 		return false
 	}
 
+	if v := normalizeVersion(f.OlderThanVersion); semver.IsValid(v) {
+		// note: invalid versions sort lower than valid versions, so an instance with
+		// an invalid version will be included in the results of queries that don't
+		// contain a "NewerThan" directive. This is intended behavior, since the primary
+		// use-case of finding old instances it to upgrade them, and any instance advertising
+		// an invalid version probably needs upgrading.
+		if semver.Compare(normalizeVersion(i.GetTeleportVersion()), v) >= 0 {
+			return false
+		}
+	}
+
+	if v := normalizeVersion(f.NewerThanVersion); semver.IsValid(v) {
+		if semver.Compare(normalizeVersion(i.GetTeleportVersion()), v) <= 0 {
+			return false
+		}
+	}
+
 	// if Services was specified, ensure instance has at least one of the listed services.
 	if len(f.Services) != 0 && slices.IndexFunc(f.Services, i.HasService) == -1 {
 		return false
 	}
 
+	if f.ExternalUpgrader != "" && f.ExternalUpgrader != i.GetExternalUpgrader() {
+		return false
+	}
+
+	// empty upgrader matches all, so we have a separate bool flag for
+	// specifically matching isntances with no ext upgrader defined.
+	if f.NoExtUpgrader && i.GetExternalUpgrader() != "" {
+		return false
+	}
+
 	return true
 }
 
+func normalizeVersion(v string) string {
```

Comment:
this is carried over from the `NormalizeVersion` helper in `lib/versioncontrol`.  I find it useful to have this just return the previous value unmodified, since it lets us sprinkle `NormalizeVersion` around in lots of places easily.